torch/autograd/_functions/tensor.py

```diff
@@ -601,13 +611,168 @@ def forward(self, input):
+
+        The second term (indexed by j) is the cumulative product of
+        x_{k+1}, x_{k+2}, ..., x_n, and it's named in the code
+        prods_from_k_pkus_1, and it's calculated as a cumprod.
+
+        In order to vectorize this properly, we need to add to
+        omitted_products the dimensions where k > j, and therefore
+        dy_j / dx_k = 0, which is done right after the assert.
+        '''
+
+        input, = self.saved_tensors
+        dim_size = input.size(self.dim)
+        if dim_size == 1:
+            return grad_output.clone()
+
+        #  Simple case with nonzero elements in the input
+        if (input != 0).all():
+            output = torch.cumprod(input, dim=self.dim)
+            return reverse_cumsum(output * grad_output, dim=self.dim) / input
+
+        if input.is_cuda:
+            idx = torch.cuda.LongTensor(range(dim_size))
+        else:
+            idx = torch.LongTensor(range(dim_size))
+
+        ones = input.select(self.dim, 0).unsqueeze(self.dim).clone().fill_(1)
+        zeros = ones.clone().fill_(0)
+        grad_input = grad_output.new(input.size()).zero_()
+        for k in range(dim_size):
+            if k == 0:
+                prods_from_k_plus_1 = torch.cumprod(
+                    input.index_select(self.dim, idx[k + 1:dim_size]),
+                    dim=self.dim
+                )
+
+                ommitted_products = torch.cat(
+                    (ones, prods_from_k_plus_1),
+                    self.dim
+                )
+
+            elif k == dim_size - 1:
+                prods_until_k = torch.prod(
+                    input.index_select(self.dim, idx[:k]),
+                    dim=self.dim
+                )
+
+                ommitted_products = prods_until_k
+
+            else:
+                prods_until_k = torch.prod(
+                    input.index_select(self.dim, idx[:k]),
+                    dim=self.dim
+                )
+
+                prods_from_k_plus_1 = torch.cumprod(
+                    input.index_select(self.dim, idx[k + 1:dim_size]),
+                    dim=self.dim
+                )
+
+                ommitted_products = prods_until_k.expand_as(
+                    prods_from_k_plus_1) * prods_from_k_plus_1
+
+                ommitted_products = torch.cat(
+                    (prods_until_k, ommitted_products), self.dim)
+
+            # At this point ommitted_products is the same size
+            # as input, except on the dimension dim where it's
+            # dim_size - k
+            assert ommitted_products.size(self.dim) == dim_size - k
+
+            if k != 0:
+                size_to_expand = list(input.size())
+                size_to_expand[self.dim] = k  # Adding zeros to missing dimensions
+                size_to_expand = torch.Size(size_to_expand)
+                ommitted_products = torch.cat(
+                    (zeros.expand(size_to_expand), ommitted_products),
+                    self.dim
+                )
```

Comment:
Maybe it would be better to just slice the grad_output, Instead of adding zero padding and doing an element-wise mul